tests/test_create_layout.py

```diff
@@ -0,0 +1,206 @@
+        {'bar/bat/four.tgz', 'foo/two.tgz', 'one.tgz', 'three.txt'})
+
+
+  def test_fully_modified_filesystem_snapshot(self):
+
+    after = {
+      'one.tgz': '1234567890aabbcc',
+      'foo/two.tgz': '0000001111112233',
+      'three.txt': '1111222233334455',
+      'bar/bat/four.tgz': '6677889900123456'
+    }
+
+    unchanged_artifacts, modified_artifacts, added_artifacts, deleted_artifacts = \
+        create_layout.changes_between_snapshots(self.before, after)
+
+    self.assertSetEqual(unchanged_artifacts, self.empty_set)
+    self.assertSetEqual(modified_artifacts,
+        {'bar/bat/four.tgz', 'foo/two.tgz', 'one.tgz', 'three.txt'})
+    self.assertSetEqual(added_artifacts, self.empty_set)
+    self.assertSetEqual(deleted_artifacts, self.empty_set)
+
+
+  def test_partially_modified_filesystem_snapshot(self):
+
+    after = {
+      'five.txt': '5555555555555555',
+      'one.tgz': '1234567890abcdef',
+      'foo/two.tgz': 'ffffffffffffffff',
+      'bar/bat/four.tgz': '6677889900123456',
+      'baz/six.tgz': '6666666666666666'
+    }
+
+    unchanged_artifacts, modified_artifacts, added_artifacts, deleted_artifacts \
+        = create_layout.changes_between_snapshots(self.before, after)
+
+    self.assertSetEqual(unchanged_artifacts, {'one.tgz'})
+    self.assertSetEqual(modified_artifacts, {'bar/bat/four.tgz', 'foo/two.tgz'})
+    self.assertSetEqual(added_artifacts, {'baz/six.tgz', 'five.txt'})
+    self.assertSetEqual(deleted_artifacts, {'three.txt'})
+
+
+  def test_create_material_rules_of_initial_step(self):
+    # Zero index means that the current step is the initial step,
+    # so we need to ALLOW all the existing files instead of matching.
+    second_link = in_toto.models.link.Link.read(self.second_step_link_str)
+    links = [second_link]
+
+    expected_materials = [
+      ['ALLOW', 'bar/bat/four.tgz'],
+      ['ALLOW', 'one.tgz'],
+      ['ALLOW', 'foo/two.tgz'],
+      ['ALLOW', 'three.txt'],
+      ['DELETE', 'three.txt'],
+      ['DISALLOW', '*']
+    ]
+
+    # create_material_rules returns a lists with random order,
+    # so we need to sort the lists
+    self.assertEqual(sorted(expected_materials),
+        sorted(create_layout.create_material_rules(None, second_link)))
+
+  def test_create_material_rules_of_not_initial_step(self):
+    # Nonzero index means that the current step is not the initial step,
+    # so we need to MATCH materials with products of the previous step.
+    first_link = in_toto.models.link.Link.read(self.first_step_link_str)
+    second_link = in_toto.models.link.Link.read(self.second_step_link_str)
+    links = [first_link, second_link]
+
+    expected_materials = [
+      ['MATCH', 'bar/bat/four.tgz', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
+      ['MATCH', 'one.tgz', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
+      ['MATCH', 'foo/two.tgz', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
+      ['MATCH', 'three.txt', 'WITH', 'PRODUCTS', 'FROM', 'first_step'],
```

Comment:
Thanks again for pointing this interesting problem out! 
For this problem, I believe the DELETE rule is more important. I think a malicious artifact won't cause many problems if we are going to delete it. But I also believe it is important to let the user know that something wrong happened in the supply chain if there is a mismatch of the artifact.

Reply:
So you say, if an artifact gets deleted in one step, it doesn't really matter if it matched the product from a previous step? That's an interesting thought, and I do agree with you on a practical level. 

However, I still think it is a lot more important to guarantee the integrity of artifacts in transit between steps than within a step. Because I trust the functionary more than the man in the middle. So I would say, match rules across steps should always be prioritized. 

Moreover, a `DISALLOW *` in the products rules of a step could catch any artifact that was expected to be deleted but wasn't actually deleted.

@SolidifiedRay, would you mind filing an issue on [in-toto/docs](https://github.com/in-toto/docs/issues) for this, and see if you can find other such conflicting rules too? Let's discuss this with other in-toto folks there before make a decision here. In the meanwhile we can just warn the caller about these cases and suggest to manually review.

Reply:
Sure! Here is the link: [issue#41](https://github.com/in-toto/docs/issues/41).

Reply:
Excellent write-up of the issue. Thanks, a lot!